Production/python/customiseHLT.py

```diff
@@ -0,0 +1,158 @@
+import FWCore.ParameterSet.Config as cms
+from PhysicsTools.NanoAOD.common_cff import Var, P4Vars
+
+def customiseGenParticles(process):
+  def pdgOR(pdgs):
+    abs_pdgs = [ f'abs(pdgId) == {pdg}' for pdg in pdgs ]
+    return '( ' + ' || '.join(abs_pdgs) + ' )'
+
+  leptons = pdgOR([ 11, 13, 15 ])
+  important_particles = pdgOR([ 6, 23, 24, 25, 35, 39, 9990012, 9900012 ])
+  process.finalGenParticles.select = [
+    'drop *',
+    'keep++ statusFlags().isLastCopy() && ' + leptons,
+    '+keep statusFlags().isFirstCopy() && ' + leptons,
+    'keep+ statusFlags().isLastCopy() && ' + important_particles,
+    '+keep statusFlags().isFirstCopy() && ' + important_particles,
+    "drop abs(pdgId) == 2212 && abs(pz) > 1000", #drop LHC protons accidentally added by previous keeps
+  ]
+
+  for coord in [ 'x', 'y', 'z']:
+    setattr(process.genParticleTable.variables, 'v'+ coord,
+            Var(f'vertex().{coord}', float, precision=10,
+                doc=f'{coord} coordinate of the gen particle production vertex'))
+  process.genParticleTable.variables.mass.expr = cms.string('mass')
+  process.genParticleTable.variables.mass.doc = cms.string('mass')
+
+  return process
+
+
+def customise(process):
+  process.NANOAODSIMoutput = cms.OutputModule("NanoAODOutputModule",
+      compressionAlgorithm = cms.untracked.string('LZMA'),
+      compressionLevel = cms.untracked.int32(9),
+      dataset = cms.untracked.PSet(
+          dataTier = cms.untracked.string('NANOAODSIM'),
+          filterName = cms.untracked.string('')
+      ),
+      fileName = cms.untracked.string('file:nano.root'),
+      outputCommands  = cms.untracked.vstring(
+          'drop *',
+          'keep nanoaodFlatTable_*Table_*_*',
+      )
+  )
+  process.load('PhysicsTools.NanoAOD.nano_cff')
+  from PhysicsTools.NanoAOD.nano_cff import nanoAOD_customizeMC
+
+  #call to customisation function nanoAOD_customizeMC imported from PhysicsTools.NanoAOD.nano_cff
+  process = nanoAOD_customizeMC(process)
+
+  process.nanoAOD_step = cms.Path(process.HLTBeginSequence \
+    + process.HLTL2TauTagNNSequence \
+    + process.HLTGlobalPFTauHPSSequence \
+    + process.HLTHPSDeepTauPFTauSequenceForVBFIsoTau \
+    + process.nanoSequenceMC)
+  process.NANOAODSIMoutput_step = cms.EndPath(process.NANOAODSIMoutput)
+
+  process.tauTable = cms.EDProducer("SimpleCandidateFlatTableProducer",
+    src = cms.InputTag( "hltHpsPFTauProducer" ),
+    cut = cms.string(""),
+    name= cms.string("Tau"),
+    doc = cms.string("HLT taus"),
+    singleton = cms.bool(False), # the number of entries is variable
+    extension = cms.bool(False), # this is the main table
+    variables = cms.PSet(
+      P4Vars,
+      charge = Var("charge", int, doc="electric charge"),
+    )
+  )
+  # cms.EDProducer : an object that produces a new data object
+
+  process.pfCandTable = cms.EDProducer("SimpleCandidateFlatTableProducer",
+    src = cms.InputTag( "hltParticleFlowForTaus" ),
```

Comment:
I wanted to ask: does this collection contain only pfCands that are part of the HLT TAU, or all pfCands in cone? i.e: would you need to add the flags that this is the part HLT TAU object?

Reply:
it contains **all** pf candidates in the event. For the training we can select all pf cands with `deltaR < X` HPS tau or AK4 jet (if we decide to not use HPS) or L1 tau seed